### PR TITLE
Controlled TextInput doesn't work most of the time

### LINKED ISSUE #7638: TextInput does not match update until it receives focus
@chrisglein Okay so I created two fresh inits, one for 0.63 and one for 0.64. I replaced the App.tsx content with the following:

```javascript
const App = () => {
  const [quantity, setQuantity] = React.useState<number>(0);
  return (
    <>
      <StatusBar barStyle="dark-content" />
      <SafeAreaView>
        <TextInput value={quantity.toString()} />
        <Button
          title={'Count Up'}
          onPress={() => {
            setQuantity(quantity + 1);
          }}
        />
      </SafeAreaView>
    </>
  );
};
```

I additionally recorded 2 small video showcases (mp4)  of the problem and added it to Google-Drive. You can watch it here:
* https://drive.google.com/file/d/1LRll1O-SYo8X8vj6QinKONegDNVdFAOt/view?usp=sharing
* https://drive.google.com/file/d/1ziNzS4bDwlwSFKP3ux1etNzGUXQkLs-s/view?usp=sharing

As you can see, the value shown by the textInput is not what it actually has registered as state. As soon as I type into the textInput it replaces the value with the correct one. After that, the count up button works.

This does definitely not align with iOS/Android/Web behaviour 🤷

_Originally posted by @creambyemute in https://github.com/microsoft/react-native-windows/issues/4444#issuecomment-824661013_

### PR DESCRIPTION
**Pull request by acoates-ms**
Fixes #7638.

JS was not providing the mostRecentEventCount back to native.  So in `TextInputShadowNode::SetText`, this condition would never be true, `m_mostRecentEventCount == m_nativeEventCount`, so any updated text value coming from JS would always be ignored.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7658)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp [hunk 1]
```diff
@@ -133,6 +133,7 @@ class TextInputShadowNode : public ShadowNodeBase {
   void SetSelection(int64_t start, int64_t end);
   winrt::Shape FindCaret(xaml::DependencyObject element);
 
+  bool m_initialUpdateComplete = false;
```
Comment (rectified95) [inline vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:136]: When is this set to true?

**Review by dannyvv — APPROVED** [review]

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp (hunk 1 above)
Comment (asklar) [inline vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:136]: consider making all these bools bits ( `: 1` ) and putting them all next to each other to improve layout

**Review by asklar — APPROVED** [review]

**Review by NickGerleman — APPROVED** [review]

**Comment by NickGerleman** [comment]
Wow this one is nasty...

Are we still actively investing in adding UI tests to RNTester examples?

**Comment by rectified95** [comment]
@NickGerleman Yes, it's still on my to-do list.

**Comment by rectified95** [comment]
@acoates-ms the `m_mostRecentEventCount == m_nativeEventCount` condition currently works for the live rewrite scenario in RNTester, 
```js
<TextInput
  onChangeText={text => {
    text = text.replace(/ /g, '_');
    this.setState({text});
  }}
  value={this.state.text}
/>
```
where `m_mostRecentEventCount` gets updated due to the `setTextAndSelection` command coming from JS. Can you check this still works after this change?

**Comment by acoates-ms** [comment]
@rectified95  yes that example seems to work with these changes.

**Comment by NickGerleman** [comment]
> @NickGerleman Yes, it's still on my to-do list.

I think test coverage will have to be a bit of a group effort. We just have too much untested code to do otherwise.

I think it would be valuable if we started adding coverage when fixing native control bugs. I think we're at the point where the framework should allow testing the sort of scenario this PR addresses without too much burden.

**Comment by NickGerleman** [comment]
OTOH there's definitely been times in the past where I felt like adding too scope (adding tests) would make a fix take too long for me to take on.

Maybe we can do something to reduce that initial barrier.

**Pull request merged**